src/ddspdrum/parameter.py

```diff
@@ -0,0 +1,159 @@
+"""
+Parameters for DDSP Modules
+"""
+
+import torch
+import torch.tensor as T
+import torch.nn as nn
+
+
+class ParameterRange:
+    """
+    ParameterRange class is a structure for keeping track of the specific range that a
+    parameter might take on. Also handles functionality for converting to and from a
+    range between 0 and 1. This class does not store the value of a parameter, just the
+    range.
+
+    Parameters
+    ----------
+    minimum (float) :   minimum value in range
+    maximum (float) :   maximum value in range
+    curve   (str)   :   relationship between parameter values and the normalized values
+                        in the range [0,1]. Must be one of "linear", "log", or "exp".
+                        Defaults to "linear"
+    """
+
+    def __init__(
+        self,
+        minimum: float = 0.0,
+        maximum: float = 1.0,
+        curve: str = "linear",
+    ):
+        self.minimum = minimum
+        self.maximum = maximum
+
+        self.curve_type = curve
+        if curve == "linear":
+            self.curve = 1
+        elif curve == "log":
+            self.curve = 0.5
+        elif curve == "exp":
+            self.curve = 2.0
+        else:
+            curve_types = ["linear", "log", "exp"]
+            raise ValueError("Curve must be one of {}".format(", ".join(curve_types)))
+
+    def __repr__(self):
+        return "ParameterRange(min={}, max={}, curve={})".format(
+            self.minimum, self.maximum, self.curve_type
+        )
+
+    def from_0to1(self, value: T) -> T:
+        """
+        Set value of this parameter using a normalized value in the range [0,1]
+
+        Parameters
+        ----------
+        value (float)   : value within [0,1] range to convert to range defined by
+            minimum and maximum
+        """
+        assert torch.all(0.0 <= value)
+        assert torch.all(value <= 1.0)
```

Comment:
@turian how should we handle values that are beyond this range? I was doing some tests with optimization and was getting errors because the parameters were going beyond this.

Reply:
I am not 100% sure.

As I wrote up earlier, I think what we want to do during training is allow to be epsilon (1e-1?) outside of [0, 1] without any penalty. And once it's smaller than -eps or greather than 1+eps, there is a strong penalty like 10 * amount outside range.

It's not entirely clear to me yet whether we should allow things outside the range here or not, or how to implement it. We can chat a little bit about it but this is fine for now and we can fix it later if we decide something else